de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java

```diff
@@ -0,0 +1,537 @@
+     *             already added
+     */
+    public synchronized void addReferencePoint(String id,
+        IReferencePoint referencePoint, boolean isPartially) {
+
+        CheckIfReferencePointIsAbleToBeShared(id, referencePoint);
+        if (isPartially) {
+            addReferencePointPartially(id, referencePoint);
+        } else if (partiallySharedReferencePoints.contains(referencePoint)) {
+            upgradeToCompletellyShared(id, referencePoint);
+            return;
+        } else {
+            addReferencePointCompletelly(id, referencePoint);
+        }
+
+        assert Collections.disjoint(completelySharedReferencePoints,
+            partiallySharedReferencePoints);
+
+        idToReferencePointMapping.put(id, referencePoint);
+        referencePointToIDMapping.put(referencePoint, id);
+
+        LOG.debug("added referencePoint " + referencePoint + " with ID " + id
+            + " [completely shared:" + !isPartially + "]");
+    }
+
+    private synchronized void CheckIfReferencePointIsAbleToBeShared(String id,
+        IReferencePoint referencePoint) {
+        if (id == null)
+            throw new NullPointerException("ID is null");
+
+        if (referencePoint == null)
+            throw new NullPointerException("referencePoint is null");
+
+        String currentReferencePointID = referencePointToIDMapping
+            .get(referencePoint);
+        IReferencePoint currentReferencePoint = idToReferencePointMapping
+            .get(id);
+
+        if (currentReferencePointID != null
+            && !id.equals(currentReferencePointID)) {
+            throw new IllegalStateException("cannot assign ID " + id
+                + " to referencePoint " + referencePoint
+                + " because it is already registered with ID "
+                + currentReferencePointID);
+        }
+
+        if (currentReferencePoint != null
+            && !referencePoint.equals(currentReferencePoint)) {
+            throw new IllegalStateException("ID " + id + " for referencePoint "
+                + referencePoint + " is already used by referencePoint "
+                + currentReferencePoint);
+        }
+    }
+
+    private synchronized void addReferencePointPartially(String id,
+        IReferencePoint referencePoint) {
+
+        if (partiallySharedReferencePoints.contains(referencePoint))
+            throw new IllegalStateException("referencePoint " + referencePoint
+                + " is already partially shared");
+
+        if (completelySharedReferencePoints.contains(referencePoint))
+            throw new IllegalStateException("referencePoint " + referencePoint
+                + " is already completely shared");
+
+        partiallySharedReferencePoints.add(referencePoint);
+
+        partiallySharedResourceMapping.put(referencePoint,
+            new HashSet<IResource>());
+    }
+
+    private synchronized void addReferencePointCompletelly(String id,
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unused method parameter "id". ([squid:S1172](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1172))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=211&issues=AWVYURY9iZQY_aXE9GJT&open=AWVYURY9iZQY_aXE9GJT)

[ISSUE_KEY]: AWVYURY9iZQY_aXE9GJT